File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/Range.java
package com.wesleyelliott.kubwa.annotation;

import com.wesleyelliott.kubwa.rule.RangeRule;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.ANNOTATION_TYPE;
import static java.lang.annotation.ElementType.CONSTRUCTOR;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.ElementType.PARAMETER;
import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(RangeRule.class)
@Retention(RUNTIME)
@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER , TYPE})
public @interface Range {
    int errorMessage();
    String name() default "rangeError";
    int min() default Integer.MIN_VALUE;
    int max() default Integer.MAX_VALUE;
    boolean includeBounds() default false;

    @Target(TYPE)
    @Retention(RUNTIME)
    @Documented
    public @interface List {
        Range[] value();
    }
}

File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/annotation/Range.kt
package com.wesleyelliott.kubwa.annotation

import com.wesleyelliott.kubwa.rule.RangeRule

/**
 * Created by wesley on 2016/07/28.
 */
@ValidateUsing(RangeRule::class)
@Retention(AnnotationRetention.RUNTIME)
@Target(
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.FIELD,
    AnnotationTarget.ANNOTATION_CLASS,
    AnnotationTarget.CONSTRUCTOR,
    AnnotationTarget.VALUE_PARAMETER,
    AnnotationTarget.CLASS,
    AnnotationTarget.FILE
)
annotation class Range(
    val errorMessage: Int,
    val name: String = "rangeError",
    val min: Int = Integer.MIN_VALUE,
    val max: Int = Integer.MAX_VALUE,
    val includeBounds: Boolean = false
) {

    @Target(AnnotationTarget.CLASS, AnnotationTarget.FILE)
    @Retention(AnnotationRetention.RUNTIME)
    @MustBeDocumented
    annotation class List(vararg val value: Range)
}
